### PR TITLE
Iris & Marj 

### PR DESCRIPTION
**Pull request by iris-lux**
# Assignment Submission: Adagrams

Congratulations! You're submitting your assignment. Please reflect on the assignment with these questions.

## Reflection

Feature | Feedback
--- | ---
What are the components that make up a method? | Name, signature, block, arguments, parameters and return value.
What are the advantages of using git when collaboratively working on one code base? | Allows a shared directory that both people can save to and pull from while using separate computers.
What kind of relationship did you and your pair have with the unit tests? | Tests for uses_available_letters? were limited in their helpfulness. There were some edge cases that they didn't account for, so bugs in that method were not found using the tests. We had to write our own unit tests for those edge cases and for is_in_english_dict?
Does your code use any methods from the `Enumerable` mixin? If so, where and why was it helpful? | We used .minmax_by in the  highest_score_from() method. It was helpful because it solved our problems with which tie to return. 
What was one method you and your pair used to debug code? | We mostly used the ruby mine debugger. We also used p - we ran some test data through the methods and printed the results. 
What are two discussion points that you and your pair discussed when giving/receiving feedback from each other that you would be willing to share? | We talked about the importance of consistently talking through our process while we were coding. We thought it was important to establish at the beginning of each coding session if it would be better to talk through the whole process or have some silent time to allow for better concentration, and then come back and discuss the code.  In general, we want to improve how to talk about our code in clear and digestible ways. We also talked about the importance of having breaks to avoid burnout and headaches.

### CODE REVIEW

File: test/adagrams_test.rb [hunk 1]
```diff
@@ -91,91 +109,27 @@
...
-    end
-
-    it 'in case of tied score and same length words, prefers the first word' do
-      # the character 'A' is worth 1 point, 'E' is 1 point
-      words = ['AAAAAAAAAA', 'EEEEEEEEEE']
-
-      # verify both have a score of 10
-      expect(score_word(words.first)).must_equal 18
-      expect(score_word(words.last)).must_equal 18
+    it 'returns false if word is not in the English dictionary' do
```
Comment (beccaelenzil) [inline test/adagrams_test.rb:124]: It looks like you removed the `highest_core_from` tests.

Comment (beccaelenzil) [reply test/adagrams_test.rb:124]: but I ran the tests, and they all pass :)

File: lib/adagrams.rb [hunk 2]
```diff
@@ -0,0 +1,141 @@
+require 'csv'
+
+
+def draw_letters
+  hand = Array.new(10)
+
+  letter_pool = [
+      {letter: "A", count: 9},
```
Comment (beccaelenzil) [inline lib/adagrams.rb:8]: Consider using the letter as the key, and the count as the value.

File: lib/adagrams.rb [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+    hand[i] = letter_pool[letter_index][:letter]
+    letter_pool[letter_index][:count] -= 1
+  end
+
+  return hand
+end
+
+def uses_available_letters?(input, letters_in_hand)
+  raise ArgumentError.new("input is nil") if input.nil?
```
Comment (beccaelenzil) [inline lib/adagrams.rb:51]: Great use of guard clauses

File: lib/adagrams.rb [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+  return uses_available
+end
+
+def score_word(word)
+
+  raise ArgumentError.new("words is nil") if word.nil?
+
+  total_score = 0
+
+  word.upcase.each_char do |letter|
```
Comment (beccaelenzil) [inline lib/adagrams.rb:87]: While this `case` statement works, it means that the information about which letter has which score is locked into this piece of code, and can't easily be used elsewhere. For example, if you wanted to display the value of each letter in a hand, you would need to repeat this work.

An alternative approach would be to store the letter scores in a hash, something like this:

```ruby
LETTER_SCORES = {
  "A" => 1
  "B" => 3,
  "C" => 3,
  "D" => 2,
  # ...
}

File: lib/adagrams.rb [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+  end
+
+  total_score += 8 if word.length > 6 && word.length <= 10
+
+  return total_score
+end
+
+def highest_score_from(words)
+
+  raise ArgumentError.new("words is nil") if words.nil?
```
Comment (beccaelenzil) [inline lib/adagrams.rb:113]: Great use of enumerable methods. You might consider breaking off lines 119 and 121 into their own tiebreaker helper method.

**Review by beccaelenzil — COMMENTED** [review]
# Adagrams



## Major Learning Goals/Code Review



| Criteria | yes/no, and optionally any details/lines of code to reference |
| --- | --- |
| Correctly creates and calls methods with proper syntax (parameters, return statements, etc.) | ✔️
| Uses correct syntax for conditional logic and iteration | ✔️
| Practices git with at least 3 small commits and meaningful commit messages | ✔️
| Utilizes unit tests to verify code; tests can run using the command `$ rake` and we see test successes and/or failures | ✔️
| Practices pair programming; the reflection question on pair programming is answered | ✔️

## Functional Requirements

| Functional Requirement | yes/no |
| --- | --- |
| For the `draw_letters` method, there is an appropriate data structure to store the letter distribution. (You are more likely to draw an 'E' than an 'X'.) | ✔️
| Utilizes unit tests to verify code; all tests for `draw_letters` and `uses_available_letters?` pass | ✔️
| Utilizes unit tests to verify code; all tests for `score_word` pass | ✔️
| Utilizes unit tests to verify code; all tests for `highest_score_from` pass | ✔️

## Overall Feedback

Great work on this project! I've left a few inline comment for you to review, but overall it is clear that the learning goals around manipulating data and writing methods were met. Your code is clear and readable, and you've made great use of enumerable methods. Keep up the hard work!

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 4+ in Code Review && 3+ in Functional Requirements | ✔️
| Yellow (Approaches Standards) | 3+ in Code Review && 2+ in Functional Requirements | ✔
| Red (Not at Standard) | 0-2 in Code Review or 0,1 in Functional Reqs, or assignment is breaking/doesn’t run with less than 5 minutes of debugging | ✔





## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation | ✅
| Elegant/Clever | ✅
| Descriptive/Readable | ✅
| Concise | ✅
| Logical/Organized | ✅
